### PR TITLE
render/gles2: use glGetAttribLocation instead of hardcoded indices

### LINKED ISSUE #2265: render/gles2: use glGetAttribLocation instead of harded indices
Right now we have magic indices for shader attribs. Instead, use `glGetAttribLocation` to retrieve the indices, so that we don't blow up when modifying the shader.

Ideally we would do one `glGetAttribLocation` call on init after compiling shaders and stash the result in the renderer struct, so that we don't need to query it each time we need to render a quad.

### PR DESCRIPTION
**Pull request by Bl4ckb0ne**
Fixes #2265

### CODE REVIEW

File: render/gles2/renderer.c [hunk 1]
```diff
@@ -205,7 +210,20 @@ static void gles2_render_ellipse_with_matrix(struct wlr_renderer *wlr_renderer,
...
-	draw_quad();
+
+
+	glVertexAttribPointer(renderer->shaders.ellipse.pos_loc, 2, GL_FLOAT,
+			GL_FALSE, 0, verts);
+	glVertexAttribPointer(renderer->shaders.ellipse.tex_loc, 2, GL_FLOAT,
+			GL_FALSE, 0, texcoord);
+
+	glEnableVertexAttribArray(0);
+	glEnableVertexAttribArray(1);
```
Comment (ascent12) [inline render/gles2/renderer.c:221]: Shouldn't these use the *_loc members too?

File: include/render/gles2.h [hunk 2]
```diff
@@ -56,11 +58,15 @@ struct wlr_gles2_renderer {
...
 			GLint color;
+			GLint pos_loc;
+			GLint tex_loc;
 		} quad;
 		struct {
 			GLuint program;
 			GLint proj;
 			GLint color;
+			GLint pos_loc;
+			GLint tex_loc;
```
Comment (emersion) [inline include/render/gles2.h:69]: I don't think the ellipse shader uses texcoords

File: render/gles2/shaders.c [hunk 3]
```diff
@@ -6,14 +6,14 @@ const GLchar quad_vertex_src[] =
 "uniform mat3 proj;\n"
 "uniform vec4 color;\n"
 "attribute vec2 pos;\n"
-"attribute vec2 texcoord;\n"
+//"attribute vec2 texcoord;\n"
```
Comment (emersion) [inline render/gles2/shaders.c:9]: We can probably get rid of this (and the rest of the texcoord stuff) for colored quads

File: render/gles2/renderer.c (hunk 1 above)
Comment (Bl4ckb0ne) [reply render/gles2/renderer.c:221]: It could yes, do you want me to change it?

File: include/render/gles2.h (hunk 2 above)
Comment (Bl4ckb0ne) [reply include/render/gles2.h:69]: The ellipse shader uses [`ellipse_fragment_src`](https://github.com/swaywm/wlroots/blob/master/render/gles2/renderer.c#L720), which uses the [`texcoord`](https://github.com/swaywm/wlroots/blob/master/render/gles2/shaders.c#L35) attribute

File: render/gles2/shaders.c (hunk 3 above)
Comment (Bl4ckb0ne) [reply render/gles2/shaders.c:9]: Oops, this was from a debug session. But yes, the `pos` vertices and the `texcoord` vertices are the same, I think we could reuse the same array of vertices.

File: render/gles2/renderer.c (hunk 1 above)
Comment (ascent12) [reply render/gles2/renderer.c:221]: Yes.

File: include/render/gles2.h (hunk 2 above)
Comment (emersion) [reply include/render/gles2.h:69]: Ah, indeed.

File: render/gles2/shaders.c (hunk 3 above)
Comment (Bl4ckb0ne) [reply render/gles2/shaders.c]: This has been put back.

File: render/gles2/renderer.c (hunk 1 above)
Comment (Bl4ckb0ne) [reply render/gles2/renderer.c]: Done for all calls to `glEnableVertexAttribArray`

File: include/render/gles2.h [hunk 4]
```diff
@@ -56,11 +58,15 @@ struct wlr_gles2_renderer {
...
```
Comment (emersion) [inline include/render/gles2.h:62]: This isn't used

File: render/gles2/renderer.c [hunk 5]
```diff
@@ -205,7 +210,20 @@ static void gles2_render_ellipse_with_matrix(struct wlr_renderer *wlr_renderer,
 
 	glUniformMatrix3fv(renderer->shaders.ellipse.proj, 1, GL_FALSE, transposition);
 	glUniform4f(renderer->shaders.ellipse.color, color[0], color[1], color[2], color[3]);
-	draw_quad();
+
+
```
Comment (emersion) [inline render/gles2/renderer.c:214]: Style: stray newline here

File: render/gles2/renderer.c [hunk 6]
```diff
@@ -86,32 +99,6 @@ static void gles2_scissor(struct wlr_renderer *wlr_renderer,
 	POP_GLES2_DEBUG;
 }
 
-static void draw_quad(void) {
```
Comment (emersion) [inline render/gles2/renderer.c:89]: I wonder if it's worth it to keep this function and make it take `pos_loc` and `tex_loc`.

Comment (emersion) [reply render/gles2/renderer.c:89]: Eh, probably not

File: render/gles2/renderer.c [hunk 7]
```diff
@@ -656,6 +678,8 @@ struct wlr_renderer *wlr_gles2_renderer_create(struct wlr_egl *egl) {
 	renderer->shaders.tex_rgbx.invert_y = glGetUniformLocation(prog, "invert_y");
 	renderer->shaders.tex_rgbx.tex = glGetUniformLocation(prog, "tex");
 	renderer->shaders.tex_rgbx.alpha = glGetUniformLocation(prog, "alpha");
+	renderer->shaders.quad.pos_loc = glGetAttribLocation(prog, "pos");
```
Comment (emersion) [inline render/gles2/renderer.c:681]: s/quad/tex_rgbx/

File: include/render/gles2.h [hunk 8]
```diff
@@ -37,6 +37,8 @@ struct wlr_gles2_tex_shader {
 	GLint invert_y;
 	GLint tex;
 	GLint alpha;
+	GLint pos_loc;
```
Comment (emersion) [inline include/render/gles2.h:40]: Nit: the existing members are locations too, so it's inconsistent to add a `_loc` suffix to the new members, we should drop it

Comment (Bl4ckb0ne) [reply include/render/gles2.h:40]: It's an attribute location compared to other that are uniform locations. `pos_attrib` maybe?

File: include/render/gles2.h (hunk 4 above)
Comment (Bl4ckb0ne) [reply include/render/gles2.h]: Done

File: render/gles2/renderer.c (hunk 5 above)
Comment (Bl4ckb0ne) [reply render/gles2/renderer.c]: Done

File: include/render/gles2.h (hunk 8 above)
Comment (emersion) [reply include/render/gles2.h:40]: Works for me

Comment (Bl4ckb0ne) [reply include/render/gles2.h]: Done

File: render/gles2/renderer.c (hunk 7 above)
Comment (Bl4ckb0ne) [reply render/gles2/renderer.c]: Done

**Comment by emersion** [comment]
Needs rebase.

**Comment by Bl4ckb0ne** [comment]
I rebased and squashed, but I'd like to have another review please, I had to make a few changes caused by the merge of https://github.com/swaywm/wlroots/pull/2092

**Pull request merged**